npeps/npep-230-dry-run.md

```diff
@@ -0,0 +1,62 @@
+# NPEP-95: NPEP template
+
+* Issue: [#230](https://github.com/kubernetes-sigs/network-policy-api/issues/230)
+* Status: Provisional
+
+## TLDR
+
+Add dry-run mode for (B)ANP to allow "disabling" policies without deleting them.
+This should allow implementations add extra audit/monitoring/logging capabilities on top of this feature.
+
+## Goals
+
+A dry-run mode should not affect any connections, and allow the networking plugin to provide feedback 
+(e.g. via logging or observability tools) to see which connections will be dropped/allowed once this (B)ANP is enforced.
+
+## Non-Goals
+
+Define exact logging/observability formats for network plugins.
+
+## Introduction
+
+Users may want to ensure no unexpected connections will be denied by a new (B)ANP.
+A dry-run mode should not affect any connections, and allow the networking plugin to provide feedback 
+(e.g. via logging or observability tools) to see which connections will be dropped/allowed once this (B)ANP is enforced.
+
+## User-Stories/Use-Cases
+
+As a cluster admin, I am designing new ANPs for my cluster and want to make sure applying them won't
+have any unexpected effects. To do so, I want to apply ANPs in a dry-run mode and get feedback from the network
+plugin on which connections would be dropped/allowed.
+
+As a cluster admin, I want to get feedback from the network plugin for currently allowed/denied connections
+with (B)ANP. 
+This one needs more discussion as it is already do-able with extra labels/annotations, because adding
+logging/audit/observability while enforcing (B)ANP is a side effect and doesn't break (B)ANP semantics.
+But since we are considering adding a new flag, it may be useful to add this option with vague description, like 
+"Enables logging/audit/observability as defined by the networking plugin. May also have no effect is the plugin doesn't support it."
+
+## API
+
+(... details, can point to PR with changes)
+
+## Conformance Details
+
+(This section describes the names to be used for the feature or
+features in conformance tests and profiles.
+
+These should be `CamelCase` names that specify the feature as
+precisely as possible, and are particularly important for
+Extended features, since they may be surfaced to users.)
+
+## Alternatives
+
+- Use Policy Assistant https://github.com/kubernetes-sigs/network-policy-api/issues/221, may have some limitations. Needs additional discussion.
+- Leave it to be implementation-specific. This required a copy of ANP CRD that changes behaviour (doesn't really apply) with dry-run flag.
+
+## References
+
+Similar features:
+- calico: https://docs.tigera.io/calico-cloud/network-policy/staged-network-policies
```

Comment:
I talked to the folks that designed Calico's staged policy and I asked why we added a second resource type `StagedNetworkPolicy` instead of an "enforced" flag.  The answer was that it allows for staging a _change_ to an existing policy as well as staging a completely new policy.  You can have a staged network policy with name "foo" and order 123 as well as an unstaged one with the same name and order.  Then the staged version should replace the active one when simulating the upcoming policy verdict.

That's not quite the same as creating a second policy at the same order with a different name; it wouldn't be clear if the user is asking us to simulate both policies being present or only the new one.

Reply:
I think this is a valuable use case, added a section about this to NPEP.
I also think it may be significantly more difficult to implement as it requires some simulation engine (at least for ovn-kubernetes), would be nice to gather more feedback for other implementations.